### ISSUE TITLE
Borg can't back up directory structure deeper than Python recursion limit

### ISSUE THREAD

**Issue by edgewood** [borgbackup/borg#380]
I noticed in the obnam 1.18 [release notes](http://blog.liw.fi/posts/obnam-1.18/) that a bug had been fixed regarding backing up directories deeper than the Python recursion limit, so I decided to see if borg was afflicted as well.  It is, and I've created a test case that currently fails.

The issue is that the _process code is recursive, and Python limits recursion to 1000 stack frames, so borg will exit with an exception if the directory hierarchy is a bit less than 1000 directories deep (_progress recurses for each nested directory, and there are half a dozen or so functions already on the stack when _progress is first called).

Test case in a branch in my fork of borg: https://github.com/edgewood/borg/tree/deep_directories

I'll look at converting _process to remove the use of recursion (maybe using os.walk()?) within a few days.

**Comment by ThomasWaldmann** [comment]
interesting, but is that really a practicaly problem?
MAX_PATH has not-that-high-limits also.

**Comment by edgewood** [comment]
Worst case we should document the limitation, or stop descending at some point and log a warning.  Ideally we should be able to back up any legal filesystem structure reachable by Python.

Let me have a few days to investigate.  If I have to bend _process into a weird shape to get it to work non-recursively, then maybe a doc note or stopping with a warning is all we should do.  But if it's still maintainable with an iterative loop...

Background about PATH_MAX:
While PATH_MAX is 4096 on Linux, I was able to create a directory hierarchy with a name 10000 characters long, stopping there because it's a round number not because that was the deepest.

$ cd /tmp; while [ $(pwd | wc -c) -lt 10000 ]; do mkdir abcde; cd abcde; done

See also the discussion in this link:

http://stackoverflow.com/questions/833291/is-there-an-equivalent-to-winapis-max-path-under-linux-unix

-- 
Ed Blackman

**Comment by ThomasWaldmann** [comment]
@edgewood oh, 10000? Interesting.

Yes, please try a iterative version, but also try to do it with minimal and easy to review changes.

**Comment by anarcat** [comment]
i'm curious to see the test case as well.

**Comment by edgewood** [comment]
On Mon, Nov 09, 2015 at 09:06:37AM -0800, anarcat wrote:
> i'm curious to see the test case as well.

Test case is in a branch of my fork of borg, 
https://github.com/edgewood/borg/tree/deep_directories

The test case is test_deep_directories in borg/testsuite/archiver.py

-- 
Ed Blackman

**Comment by edgewood** [comment]
The deep_directories branch is up-to-date with my current progress.  I've fixed the testcase to clean itself up: shutil.rmtree() calls itself recursively so it breaks like borg does when the hierarchy is too deep.  I've split the "walk the tree" code from the "process files/directories" code successfully: all the testcases except the new one pass.  The new one still doesn't pass because I used os.walk() to walk the tree, and... os.walk() calls itself recursively.  D'oh!  Apparently if you want to deal with deep directory structures in Python you have to do it yourself.

So I'll need to create a method or a class to scan the entry_path and return a list of subpaths to process.  I do have a source of inspiration, see scan_tree in http://git.liw.fi/cgi-bin/cgit/cgit.cgi/obnam/tree/obnamlib/vfs.py. It's a generator that uses a Python list as a stack of items to process, yielding items one at a time.

It also shows an interesting pattern (the ok parameter) to clean up the huge if/elif/else block (path excluded? is it in skip_inodes, should we cross FS boundaries? etc) in the current _process.  But I'll leave that for later.

Review of my current code appreciated, with the caveat that it's not done yet.

**Comment by ThomasWaldmann** [comment]
Please also have a look at py35 walk code, it was significantly sped up compared to earlier versions.

If we change this code, we maybe could also go breadth-first, which is better for fuse performance (and also better if we want to store directory content into the repo at some time).

**Comment by anarcat** [comment]
i believe the py35 `walk()` code simply switches from `listdir()` to `scandir()`, so it skips a `stat()` call which makes it faster, but is still recursive.

**Comment by ThomasWaldmann** [comment]
Maybe a good goal would be to replace os.walk by a very compatible nonrecursive function, so this could be proposed for stdlib.

**Comment by anarcat** [comment]
On 2015-11-12 20:41:32, TW wrote:
> Maybe a good goal would be to replace os.walk by a very compatible nonrecursive function, so this could be proposed for stdlib.

sounds like a plan. :) this *is* how scandir came to be after all.

-- 
Instead of worrying about what somebody else is going to do, which is
not under your control, the important thing is, what are you going to
decide about what is under your control?
                         - Richard Stallman

**Comment by ThomasWaldmann** [comment]
Q: can we close this here and move this to python issue tracker?

As long as the python implementation of os.walk is recursive, it is maybe a bit pointless if we fix our code (which uses os.walk) to be non-recursive.

Also, I still think this has little practical relevance and if you run into this, you likely rather want to fix your filesystem.

**Comment by ThomasWaldmann** [comment]
Python tracker issue: http://bugs.python.org/issue26545

**Comment by enkore** [comment]
It shouldn't outright crash. Log + resume from RecursionError?

**Comment by ThomasWaldmann** [comment]
Tag "later", for one day when we are totally bored? :-)

**Comment by enkore** [comment]
Yeah, it's not really _that_ important...

**Comment by cmcginty** [comment]
If this becomes an issue there could be an environment option added to increase the limit.

https://docs.python.org/3.7/library/sys.html#sys.setrecursionlimit

**Comment by ThomasWaldmann** [comment]
Note: we could close this issue by:

- adding a new env var BORG_RECURSIONLIMIT (code and docs)
- checking if it is set, if yes, call sys.setrecursionlimit with int(value)

PRs welcome.

**Comment by LaurentBonnaud** [comment]
This bug still exists in borg 1.1.15.

I am adding a backtrace here because I did not find this bug by a general Web search and I think it might be useful for people affected by this bug:
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4621, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4553, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 177, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 590, in do_create
    create_inner(archive, cache)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 553, in create_inner
    read_special=args.read_special, dry_run=dry_run, st=st)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 667, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 667, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 667, in _process
    read_special=read_special, dry_run=dry_run)
  [Previous line repeated 978 more times]
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 659, in _process
    status = archive.process_dir(path, st)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 948, in process_dir
    with self.create_helper(path, st, 'd', hardlinkable=False) as (item, status, hardlinked, hardlink_master):
  File "/usr/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 930, in create_helper
    item = Item(path=safe_path)
  File "src/borg/item.pyx", line 53, in borg.item.PropDict.__init__
  File "src/borg/item.pyx", line 59, in borg.item.PropDict.update
  File "src/borg/item.pyx", line 116, in borg.item.PropDict._make_property._set
  File "/usr/lib/python3/dist-packages/borg/helpers.py", line 1023, in safe_encode
    return s.encode(coding, errors)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

**Comment by ThomasWaldmann** [comment]
@LaurentBonnaud ok, can you tell me what the limits of your fs are (esp. directory depth)?

I think I have an idea we could try: treat this error like a per-file (per directory) error.

So the exception would still happen, but get caught and processing would just skip that "too deep" directory **and continue doing the remainder of the backup**.

**Comment by ThomasWaldmann** [comment]
```
user@development:~$ python3
Python 3.7.3 (default, Jan 22 2021, 20:04:44) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.

>>> def recurse(n):
...  print(n)
...  try:
...   recurse(n+1)
...  except RecursionError:
...   print("recursion error")
...  print(n)
... 
>>> recurse(0)
0
1
2
3
...
993
994
995
recursion error
994
993
992
...
3
2
1
0
```

Seems to *almost* work, but why is there no **2nd 995** after the recursion error?

**Comment by LaurentBonnaud** [comment]
> can you tell me what the limits of your fs are

I am using ext4 and I don't know its limits precisely. My guess is that the limitation comes from PATH_MAX that is equal to 4096. So that would be roughly 2048 directories.

**Comment by jdchristensen** [comment]
I haven't looked at the code here, but can the recursion be easily removed by just adding directories to a list (or other data structure) and using a loop?  This kind of thing is often a fairly routine code transformation.

**Comment by ThomasWaldmann** [comment]
I did some effort today in changing to a non-recursive walking.

TL;DR: it's complicated and in the end, I abandoned it. Too many issues, too little to win.

Issues included:
- dealing with open FDs (borg works a lot with FDs trying to avoid race conditions when the fs changes)
- timestamps of directories when extracting them (with their content), order is important
- code gets less readable
- even if we manage stuff on a self-made stack (python list), we have to limit the stack size or it could happen that a fs defect leads to an ever growing stack and an endlessly running borg create. so we have to define some arbitrary limit and thus we are just exchanging python's limit to our limit.
- python's limit (1000) is quite high, if you have deeper nesting it is usually due to some fs defect or somebody trying to annoy you. you need to monitor your backups, not just for that reason.
